### PR TITLE
[ENGG-1588] feat: async support in response rule - XHR requests

### PR DESCRIPTION
**Pull request by wrongsahil**
Closes issue: #960 

## ✅ Checklist:

- [ ] Make sure linting and unit tests pass.
- [ ] No install/build warnings introduced.
- [ ] Verified UI in browser.
- [ ] For UI changes, added/updated analytics events (if applicable).
- [ ] For changes in extension's code, manually tested in Chrome and Firefox.
- [ ] For changes in extension's code, both MV2 and MV3 are covered.
- [ ] Added/updated unit tests for this change.
- [ ] Raised pull request to update corresponding documentation (if already exists).

### CODE REVIEW

**Comment by anlinguist** [comment]
👀  nice! this is certainly a better approach than my double request approach. Is it finished or still WIP?

**Comment by wrongsahil** [comment]
> 👀 nice! this is certainly a better approach than my double request approach. Is it finished or still WIP?

@anlinguist We are doing final round of testing. Will release this next week 😬

File: browser-extension/mv2/src/client/js/requestResponseRuleHandler.js [hunk 1]
```diff
@@ -454,128 +513,181 @@ RQ.RequestResponseRuleHandler.interceptAJAXRequests = function ({
...
-    const xhr = new XHR();
-    xhr.addEventListener("readystatechange", onReadyStateChange.bind(xhr), false);
+    const xhr = new OriginalXMLHttpRequest();
+    createProxyXHRObject.call(xhr);
     return xhr;
   };
-  XMLHttpRequest.prototype = XHR.prototype;
-  Object.entries(XHR).map(([key, val]) => {
+
+  // TODO: Why this required? Check with @nafees/vaibhav
```
[the remark this answers is not in the document]
Comment (wrongsahil) [reply browser-extension/mv2/src/client/js/requestResponseRuleHandler.js:600]: Ohh. Got it. Thanks :D

File: browser-extension/mv2/src/client/js/requestResponseRuleHandler.js [hunk 2]
```diff
@@ -454,128 +508,177 @@ RQ.RequestResponseRuleHandler.interceptAJAXRequests = function ({
...
+    this._abort = true;
+    this.rqProxyXhr._abort = true;
+    abort.apply(this, arguments);
+    abort.apply(this.rqProxyXhr, arguments);
+  };
+
+  let setRequestHeader = XMLHttpRequest.prototype.setRequestHeader;
+  XMLHttpRequest.prototype.setRequestHeader = function (header, value) {
+    this._requestHeaders = this._requestHeaders || {};
+    this._requestHeaders[header] = value;
```
Comment (lazyvab) [inline browser-extension/mv2/src/client/js/requestResponseRuleHandler.js:620]: can we remove and store only in `this.rqProxyXhr`?

File: browser-extension/mv2/src/client/js/requestResponseRuleHandler.js [hunk 3]
```diff
@@ -454,128 +508,177 @@ RQ.RequestResponseRuleHandler.interceptAJAXRequests = function ({
...
+    this.rqProxyXhr._method = method;
+    this.rqProxyXhr._requestURL = getAbsoluteUrl(url);
+    open.apply(this.rqProxyXhr, arguments);
     open.apply(this, arguments);
   };
 
+  const abort = XMLHttpRequest.prototype.abort;
+  XMLHttpRequest.prototype.abort = function () {
+    isDebugMode && console.log("abort called");
+    this._abort = true;
```
Comment (lazyvab) [inline browser-extension/mv2/src/client/js/requestResponseRuleHandler.js:611]: can we remove and store only in `this.rqProxyXhr`?

**Review by lazyvab — APPROVED** [review]

**Pull request merged**